### PR TITLE
Draft PR for adding support for custom attribution

### PR DESCRIPTION
**Pull request by jatgarg**
### E2E/Fuzz/Unit Tests are not added in this PR since this solution is in discussion.

Attribution means who inserted/annotated the text and at what time. Currently FF only supports attribution for the text and attribution for the props in the shared string. This means that the App can only get info about the user Id and the time at which a text is inserted/annotated like bold/italic or any other prop. We don’t allow app to store any other attribution data related to piece of text. For example,  

1.) App wants to add a prop called DataSource which tells whether the source of text was from a template or not (called DataSource attribution prop on shared string). 

2.) For copy paste scenario, where the user could be anyone outside the users of doc/org or even inside the doc, we need to let the app insert that custom attribution info which cannot be extracted from the op. 

3.) Some copilot attribution props like “onBehalfOf” etc which tells us the user on behald of whom the text was inserted by the copilot. 

**Solution 2 link here: https://github.com/microsoft/FluidFramework/pull/22053**

Solution:

1.) Add another type of key alongside the type of attribution keys which we already have. 
**CustomAttributionKey = {type: “custom”, id: string }** 
App will call createCustomAttributionKey() on the runtime attributor with the attribution data the user wants to store which will return custom attribution keys to the app. 
API on runtime attributor: 
**createCustomAttributorKey(
		attributionInfoList: CustomAttributionInfo[],
	): CustomAttributionKey[];**

We will take an array of attributes object, so that user does not have to call it multiple times to create multiple keys before let's say passing those keys to insertText in shared string. 
The ids in the custom attribution keys would be generated using id compressor to keep them short and if id compressor is not available, then we will fall back to uuids.

2.) App will use these keys to pass to the shared string Apis like insertText() in props object. It can pass multiple keys at once as we can have multiple keys for a piece of text at different offsets like in copy paste scenario. 
This is represented by ICustomAttributionKeyList which can be added to props by utility api like: **insertCustomAttributionPropInPropertySet()**

We will define a channel name for the AttributionCollection for the segment which will be reserved for providing the custom keys generated above. 
Let’s call it: const customAttributionKeysPropName  = “CAK-guid” 
where CAK is abbreviation for Custom Attribution Key. Just having a guid so as to avoid probable collision with other props which a user could insert and we will also match the type of this prop to make sure it is really meant for supplying custom attribution keys from the app. 
 
sharedstring.insertText(pos, text, { customAttributionKeysPropName: [{offset: 0, key: CustomAttributionKey1}, {offset: 5, Key: CustomeAttributionKey2 }]}. 
This info will be stored in the attribution collection on the main collection once it is processed and this prop will be deleted from the props once consumed.
We will only allow app to insert all this info only on insertion as this should not be changed once supplied.

3.) Transfer of above knowledge to other clients: 
We will add an API in container runtime class which will allow to submit external ops and then allow attributor to supply a external op processor so that other clients can process the external attributor op and add the custom attribution keys for query and summary later. 
**public submitExternalOps(innerType: string, contents: unknown, localOpMetadata)** 

For processing of these ops at other clients, the runtime attributor will add an op processor to the constructor.  
export interface IExternalOpProcessor {
	opProcessor: (
		op: ContainerRuntimeExternalOp,
		local: boolean,
		localOpMetadata?: unknown,
	) => boolean;
	stashedOpProcessor: (op: ContainerRuntimeExternalOp) => Promise<boolean>;
}
 

4.) Clients can then use the current api: attributor.getAtOffset(offset, cannelName) to get the custom attribution keys. Ex. 
attributor.getAtOffset (0,  customAttributionKeysPropName): CustomAttributionKey[] 

5.) Which can used to query the attributor to get the corresponding Attribution Info. 
Ex. Attributor.getCustomAttributionInfo(customAttributionKey): CustomAttributionInfo.

### CODE REVIEW

File: packages/dds/merge-tree/src/attributionCollection.ts [hunk 1]
```diff
@@ -558,3 +563,19 @@ export class AttributionCollection implements IAttributionCollection<Attribution
...
+export class CustomAttributionCollection extends AttributionCollection {
+	public constructor(_length: number, customAttributionKeyList: ICustomAttributionKeyList) {
+		super(_length);
+		const keysArr = customAttributionKeyList.keys;
+
+		// Sort the keysArr array by the offset as they might be not sorted.
+		keysArr.sort((a, b) => a.offset - b.offset);
+
+		// Now that the keysArr array is sorted, separate the keys and values back out
+		for (const [_, pair] of keysArr.entries()) {
```
Comment (vladsud) [inline packages/dds/merge-tree/src/attributionCollection.ts:576]: Why not use keysArr.values() instead?

File: packages/dds/merge-tree/src/attributionPolicy.ts [hunk 2]
```diff
@@ -188,6 +196,50 @@ function createPropertyTrackingMergeTreeCallbacks(
...
+				propName
+			] as ICustomAttributionKeyList;
+			const customCollection = new CustomAttributionCollection(
+				segment.cachedLength,
+				customAttributionKeyList,
+			);
+			segment.attribution?.update(customAttributionKeysPropName, customCollection);
+			// Delete it from the segment properties as we don't want it to get summarized twice as
+			// we have added it in segment.attribution which will be summarized.
+			delete segment.properties?.[propName];
```
Comment (vladsud) [inline packages/dds/merge-tree/src/attributionPolicy.ts:219]: delete is rather expensive in JS (usually). is it possible to use segment.properties?.[propName] = undefined instead?
It's not the same as delete.

Some quick search results (if you are curious to read more): https://stackoverflow.com/questions/59431589/javascript-performance-constraints-of-delete-keyword

File: packages/dds/merge-tree/src/attributionUtils.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+import { assert } from "@fluidframework/core-utils/internal";
+import type { CustomAttributionKey } from "@fluidframework/runtime-definitions/internal";
+
+import type { PropertySet } from "./properties.js";
+
+/**
+ * @internal
+ */
+export const customAttributionKeysPropName: string =
+	"CAK-cf9b6fe4-4c50-4a5d-9045-eb73b886f740";
```
Comment (vladsud) [inline packages/dds/merge-tree/src/attributionUtils.ts:15]: I wonder what is "CAK"? Looks like guid with some prefix - why do we need prefix?
Also, I assume we do not care about using actual guids, only (maybe) to isolate user names from "system" names. In some other places we disallow user names to start with "_" (or some other special character) and use that for short system names.

Not sure if it matters, but having long IDs on a lot of small DDSs is not great (tablero default case)

File: packages/dds/merge-tree/src/attributionUtils.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+export function isCustomAttributionKeyList(obj: unknown): obj is ICustomAttributionKeyList {
+	return (
+		obj !== null &&
+		typeof obj === "object" &&
+		"type" in obj &&
+		"keys" in obj &&
+		obj.type === "custom" &&
+		Array.isArray(obj.keys)
```
Comment (vladsud) [inline packages/dds/merge-tree/src/attributionUtils.ts:32]: I think you can simply that:

```ts
obj?.type === "custom" &&
Array.isArray(obj?.keys)
```

That said, this feels like pretty weak detection of a type system.
Should it has some unique (long) property that is tested here?
I.e. `obj?._customAttributionKeyList === true`

File: packages/dds/merge-tree/src/attributionUtils.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+ */
+export function insertCustomAttributionPropInPropertySet(
+	props: PropertySet,
+	offsets: number[],
+	keys: CustomAttributionKey[],
+): void {
+	if (keys.length === 0) {
+		return;
+	}
+	assert(offsets.length === keys.length, "offsets and keys length should be same");
```
Comment (vladsud) [inline packages/dds/merge-tree/src/attributionUtils.ts:51]: move assert above previous if/return block?

File: packages/framework/attributor/src/mixinAttributor.ts [hunk 6]
```diff
@@ -275,6 +383,9 @@ class RuntimeAttributor implements IRuntimeAttributor {
 			return false;
 		}
 
+		if (key.type === "custom") {
+			return this.opAttributor?.tryGetCustomAttributionInfo(key.id) !== undefined;
```
Comment (vladsud) [inline packages/framework/attributor/src/mixinAttributor.ts:387]: maybe instead of having "try" method, it's better to have "has" method?

File: packages/framework/attributor/src/encoders.ts [hunk 7]
```diff
@@ -108,6 +166,46 @@ export class AttributorSerializer implements IAttributorSerializer {
...
+			const {
+				ids,
+				timestamps: encodedTimestamps2,
+				attributionRefs: attributionRefs2,
+				attributionPropsRefs,
+				attributionValuesRefs,
+			} = customAttributionInfo;
+			const timestamps2 = this.timestampEncoder.decode(encodedTimestamps2);
+			assert(
+				ids.length === timestamps2.length &&
```
Comment (vladsud) [inline packages/framework/attributor/src/encoders.ts:180]: maybe cache ids.length into local variable and use it below? Complier will use single letter variable for it - a bit more compact.

File: packages/framework/attributor/src/encoders.ts [hunk 8]
```diff
@@ -108,6 +166,46 @@ export class AttributorSerializer implements IAttributorSerializer {
...
+					ids.length === attributionRefs2.length &&
+					ids.length === attributionPropsRefs.length &&
+					ids.length === attributionValuesRefs.length,
+				"serialized cutom attribution columns should have the same length",
+			);
+			const entries2: [string, CustomAttributionInfo][] = Array.from({ length: ids.length });
+			for (const [i, key] of ids.entries()) {
+				// Non null asserting, we asserted ids, timestamps and attributionRefs have the same length above
+				// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+				const timestamp = timestamps2[i]!;
```
Comment (vladsud) [inline packages/framework/attributor/src/encoders.ts:190]: Curious, why do you need to use `!` here and other places? They all seem to be plain typed arrays, and quickly checking in TS playground, it gives me actual type of the element, without adding undefined to it.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 9]
```diff
@@ -3923,6 +3944,23 @@ export class ContainerRuntime
 		this.submit({ type, contents }, localOpMetadata);
 	}
 
+	/**
+	 * Api to allow users of runtime to submit ops if required.
+	 * @param type - Type of external op.
+	 * @param contents - contents of the op.
+	 * @param localOpMetadata - localOpMetadata is any.
+	 */
+	public submitExternalOps(
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:3953]: This mechanism looks suspicious.
You essentially depend now on a chain of these processors for correctness, and there are no protections that all clients will have same code hooks installed to guarantee eventual consistency. For example code does not assert if there is no proper this.externalOpProcessors for an op.
It is also not obvious that resubmit can be simple, i.e. not require rebasing.

I wonder if what you need here is a real DDS and just let DDS (and code associated with it) handle all of that. We are missing ability to install "system" DDSs under root, but we will need to implement such capability in the future (for Tree DDS to rule the world (runtime)).

Basically, we should be very careful to add any kind of custom behavior like that to container runtime - it's already too complicated and has too many custom flows (like ID compressor, that ideally should also be a DDS).

File: packages/runtime/container-runtime/src/messageTypes.ts [hunk 10]
```diff
@@ -237,3 +256,19 @@ export type InboundSequencedContainerRuntimeMessageOrSystemMessage =
...
+	type: string;
+
+	contents: unknown;
+}
+
+/**
+ * @alpha
+ * @legacy
+ */
+export type TExternalContainerRuntimeMessage = IExternalContainerRuntimeMessage;
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/messageTypes.ts:274]: What is the reason we need this type? I.e. why IExternalContainerRuntimeMessage is not sufficient?
Can you please add some description over for these types?
